src/online2/online-nnet3-wake-word-faster-decoder.h

```diff
@@ -0,0 +1,145 @@
+  };
+
+  // "sil_phones" - the IDs of all silence phones
+  OnlineWakeWordFasterDecoder(const fst::Fst<fst::StdArc> &fst,
+                              const OnlineWakeWordFasterDecoderOpts &opts,
+                              const std::vector<int32> &sil_phones,
+                              const TransitionModel &trans_model,
+                              const nnet3::DecodableNnetSimpleLoopedInfo &info,
+                              OnlineNnet2FeaturePipeline *features)
+      : FasterDecoder(fst, opts), opts_(opts),
+        silence_set_(sil_phones), trans_model_(trans_model),
+        max_beam_(opts.beam), effective_beam_(FasterDecoder::config_.beam),
+        state_(kEndFeats), frame_(0), utt_frames_(0),
+        input_feature_frame_shift_in_seconds_(features->FrameShiftInSeconds()),
+        decodable_(trans_model_, info,
+            features->InputFeature(), features->IvectorFeature()) {}
+
+  DecodeState Decode();
+  
+  // Makes a linear graph, by tracing back from the last "immortal" token
+  // to the previous one
+  bool PartialTraceback(fst::MutableFst<LatticeArc> *out_fst);
+
+  // Makes a linear graph, by tracing back from the best currently active token
+  // to the last immortal token. This method is meant to be invoked at the end
+  // of an utterance in order to get the last chunk of the hypothesis
+  void FinishTraceBack(fst::MutableFst<LatticeArc> *fst_out);
+
+  // Returns "true" if the best current hypothesis ends with long enough silence
+  bool EndOfUtterance();
+
+  int32 frame() { return frame_; }
+
+ private:
+  void ResetDecoder(bool full);
+
+  // Returns a linear fst by tracing back the last N frames, beginning
+  // from the best current token
+  void TracebackNFrames(int32 nframes, fst::MutableFst<LatticeArc> *out_fst);
+
+  // Makes a linear "lattice", by tracing back a path delimited by two tokens
+  void MakeLattice(const Token *start,
+                   const Token *end,
+                   fst::MutableFst<LatticeArc> *out_fst) const;
+
+  // Searches for the last token, ancestor of all currently active tokens
+  void UpdateImmortalToken();
+
+  const OnlineWakeWordFasterDecoderOpts opts_;
+  const ConstIntegerSet<int32> silence_set_; // silence phones IDs
+  const TransitionModel &trans_model_; // needed for trans-id -> phone conversion
+  const BaseFloat max_beam_; // the maximum allowed beam
+  BaseFloat &effective_beam_; // the currently used beam
+  DecodeState state_; // the current state of the decoder
+  int32 frame_; // the next frame to be processed
+  int32 utt_frames_; // # frames processed from the current utterance
+  Token *immortal_tok_;      // "immortal" token means it's an ancestor of ...
+  Token *prev_immortal_tok_; // ... all currently active tokens
+  // this is remembered from the constructor; it's ultimately
+  // derived from calling FrameShiftInSeconds() on the feature pipeline.
+  BaseFloat input_feature_frame_shift_in_seconds_;
```

Comment:
I'm not sure why this frame shift is needed, but bear in mind that the nnet computation may have frame-subsampling-factor set, so you should probably get this another way if it's needed.